pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/admin/UserAdminService.java

```diff
@@ -0,0 +1,844 @@
+                    + "JOIN access_user_role_group aurg on aurg.user_id = au.user_id "
+                    + "JOIN ddp_group dg on dg.group_id = aurg.group_id "
+                    + "WHERE dg.group_id = ? AND au.is_active = 1";
+
+    private static final String SQL_INSERT_USER_ROLE =
+            "INSERT INTO access_user_role_group SET user_id = ?, role_id = ?, group_id = ?";
+
+    private static final String SQL_DELETE_USER_ROLES =
+            "DELETE FROM access_user_role_group WHERE user_id = ?";
+
+    private static final String SQL_DELETE_GROUP_ROLE =
+            "DELETE FROM ddp_group_role WHERE group_role_id = ?";
+
+    private static final String SQL_DELETE_USER_ROLE =
+            "DELETE FROM access_user_role_group WHERE user_id = ? AND role_id = ? AND group_id = ?";
+
+    private static final String SQL_INSERT_GROUP =
+            "INSERT INTO ddp_group SET name = ?";
+
+    private static final String SQL_DELETE_GROUP =
+            "DELETE FROM ddp_group WHERE group_id = ?";
+
+    private static final String SQL_INSERT_GROUP_ROLE =
+            "INSERT INTO ddp_group_role SET group_id = ?, role_id = ?, admin_role_id = ?";
+
+    private static final String SQL_SELECT_GROUP_FOR_REALM =
+            "SELECT dg.group_id, dg.name from ddp_group dg "
+                    + "JOIN ddp_instance_group dig on dig.ddp_group_id = dg.group_id "
+                    + "JOIN ddp_instance di on di.ddp_instance_id = dig.ddp_instance_id "
+                    + "WHERE di.instance_name = ?";
+
+    public UserAdminService(String operatorId, String studyGroup) {
+        this.operatorId = operatorId;
+        this.studyGroup = studyGroup;
+    }
+
+    /**
+     * Get study roles that operator can administer
+     */
+    public StudyRoleResponse getStudyRoles() {
+        int groupId = verifyStudyGroup(studyGroup);
+        Map<String, RoleInfo> studyRoles = getAdminRoles(groupId);
+
+        StudyRoleResponse res = new StudyRoleResponse();
+        for (var entry: studyRoles.entrySet()) {
+            RoleInfo roleInfo = entry.getValue();
+            res.addRole(new StudyRoleResponse.Role(roleInfo.getName(), roleInfo.getDisplayText()));
+        }
+        return res;
+    }
+
+    /**
+     * Add roles for list of users
+     */
+    public void addUserRoles(UserRoleRequest req) {
+        validateRoleRequest(req);
+
+        int groupId = verifyStudyGroup(studyGroup);
+        Map<String, RoleInfo> studyRoles = getAdminRoles(groupId);
+        List<String> roleNames = req.getRoles();
+        validateRoles(roleNames, studyRoles.keySet());
+
+        for (String userEmail: req.getUsers()) {
+            if (StringUtils.isBlank(userEmail)) {
+                throw new DSMBadRequestException("Invalid user email: blank");
+            }
+            addRoles(userEmail, roleNames, groupId, studyRoles);
+        }
+    }
+
+    protected void addRoles(String email, List<String> roles, int groupId, Map<String, RoleInfo> studyRoles) {
```

Comment:
Is this method only getting the newly added roles for the participant? To avoid adding a role multiple times for a user

Reply:
This method adds the requested roles (`roles`) for one user (`email`), and validates the roles against `studyRoles`, which also has the role ID (so we don't have to ask the DB for it again). As you noted below, this is improved in the upcoming branch to validate the requested roles before calling `addRoles`.

Reply:
I see! Thank you for explaining that!
One thing I am still confused about (sorry if it sounds too much overthinking, probably because I have not see the frontend and how it will be handled) is if user `A` already has the` kit_shipping` role, and in frontend the admin also assigns the `kit_deactivation` role for them, are both of those roles coming to `addRoles` in the `roles` List and the `kit_shipping` is then added twice? I think there should be a check for what is a "new" role vs an existing role of a user, either from frontend or backend to stops that, but I think I missed it here.

Reply:
Ah, I think I answered the wrong question initially. 
If I understand correctly, you are asking what prevents the code from adding the same role for a user twice, which is a valid concern since there is no key constraint to prevent that (and thus we have users with more than one instance of the same role, at least on Dev).
We do that check at the beginning of addUserRole, which is the only place we add user roles.

Does that answer your question, or did I misunderstand again?